pkg/tools/cli/loadtest/service/loader.go

```diff
@@ -88,48 +99,64 @@ func (l *loader) Prepare(context.Context) (err error) {
 	if err != nil {
 		return err
 	}
-	l.requests, err = l.requestsFunc(dataset)
+
+	switch l.operation {
+	case config.Insert, config.StreamInsert:
+		l.dataProvider, l.dataSize, err = insertRequestProvider(dataset, l.batchSize)
+	case config.Search, config.StreamSearch:
+		l.dataProvider, l.dataSize, err = searchRequestProvider(dataset)
+	}
 	if err != nil {
 		return err
 	}
 
-	if len(l.requests) == 0 {
-		return errors.New("prepare data is empty")
-	}
-
 	return nil
 }
 
 // Do operates load test.
 func (l *loader) Do(ctx context.Context) <-chan error {
```

Comment:
There is a part omitted, but I think it will be beautiful if you divide `Do`. :bow:

```go
// Do operates load test.
func (l *loader) Do2(ctx context.Context) <-chan error {
	ech := make(chan error, l.dataSize)

	notifyErr := func(ctx context.Context, err error) {
		select {
		case <-ctx.Done():
			ech <- errors.Wrap(err, ctx.Err().Error())
		case ech <- err:
		}
	}

	// TODO: related to #403.
	process, err := os.FindProcess(os.Getpid())
	if err != nil {
		notifyErr(ctx, err)
		return ech
	}

	ticker := time.NewTicker(l.progressDuration)
	l.eg.Go(safety.RecoverFunc(func() error {
		defer ticker.Stop()

		return nil
	}))

	l.eg.Go(safety.RecoverFunc(func() error {
		log.Infof("start load test(%s, %s)", l.service.String(), l.operation.String())
		defer close(ech)
		defer ticker.Stop()

		_, _, err := l.do(ctx, notifyErr)
		if err != nil {
			notifyErr(ctx, err)
			return process.Signal(syscall.SIGKILL)
		}

		return process.Signal(syscall.SIGTERM)
	}))

	return ech
}

func (l *loader) do(ctx context.Context, notifyErr func(context.Context, error)) (start, end time.Time, err error) {
	start = time.Now()
	defer func() {
		end = time.Now()
	}()

	eg, egctx := errgroup.New(ctx)

	switch l.operation {
	case config.StreamInsert, config.StreamSearch:
		var newData = func() interface{} {
			switch l.operation {
			case config.StreamSearch:
				return func() interface{} {
					return new(payload.Search_Response)
				}
			default:
				return func() interface{} {
					return new(payload.Empty)
				}
			}
		}

		eg.Go(safety.RecoverFunc(func() error {
			conn, err := grpc.Dial(l.addr, grpc.WithInsecure())
			if err != nil {
				notifyErr(egctx, err)
				return nil
			}
			defer notifyErr(egctx, conn.Close())

			st, err := l.loaderFunc(egctx, conn, nil)
			if err != nil {
				notifyErr(egctx, err)
				return nil
			}
			err = igrpc.BidirectionalStreamClient(st.(grpc.ClientStream), l.dataProvider, newData, nil)
			if err != nil {
				notifyErr(egctx, err)
			}
			return nil
		}))
		err = eg.Wait()

	case config.Insert, config.Search:
		eg.Limitation(l.concurrency)

		for {
			r := l.dataProvider()
			if r == nil {
				break
			}

			eg.Go(safety.RecoverFunc(func() error {
				conn, err := grpc.Dial(l.addr, grpc.WithInsecure())
				if err != nil {
					notifyErr(egctx, err)
					return nil
				}
				defer notifyErr(egctx, conn.Close())

				_, err = l.loaderFunc(egctx, conn, r)
				if err != nil {
					notifyErr(egctx, err)
				}
				return nil
			}))
		}
		err = eg.Wait()

	default:
		err = errors.Errorf("undefined type: %s", l.operation.String())
	}

	return
}
```

Reply:
Thanks a lot! I fix it now!